pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -227,6 +259,49 @@ func (meta *predicateMetadata) RemovePod(deletedPod *v1.Pod) error {
 	return nil
 }
 
+func updateAffinityOrAntiAffinityInTPM(tpm *topologyPairsMaps, terms []v1.PodAffinityTerm, node *v1.Node,
+	incomingPod, existingPod *v1.Pod, isAffinity bool) error {
+	for idx, term := range terms {
+		namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(incomingPod, &term)
+		selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
+		if err != nil {
+			return err
+		}
+		if priorityutil.PodMatchesTermsNamespaceAndSelector(existingPod, namespaces, selector) {
+			if topologyValue, ok := node.Labels[term.TopologyKey]; ok {
+				pair := topologyPair{key: term.TopologyKey, value: topologyValue}
+				if isAffinity {
+					tpm.addTopologyPair(pair, existingPod, sets.NewInt(idx))
+					tpm.podAffinityMatchMap[idx]++
+				} else {
+					tpm.addTopologyPair(pair, existingPod, sets.Int{})
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// satifiesPodAffinity calculates if podAffinity of "pod" can fit or not
+// on the whole cluster
+// note: this function is not expected to be called externally (outside metadata.go)
+// other potential callers should use meta.isPodAffinitySatisfied instead
+func satifiesPodAffinity(pod *v1.Pod, podAffinityMatchMap map[int]int) bool {
```

Comment:
@bsalamat you're right, `map[int]int` is not enough. Let me update that data structure.

Reply:
> In fact figuring out whether the node affinity can be satisfied anywhere in the cluster without actually checking all the nodes is the biggest challenge.

Agree. My current idea is to first implement a helper structure which can return all node names by searching topologyKey (a.k.a node label) in O(1) time. It's quite similar like `NodeTree`.

Then based on this, I can know which nodes fit for which affinityTerm, and doing an intersection among nodes would give the final result.
The time complexity would be O(n•log<sub>2</sub>m)
- n means length of node matching a specific affinityTerm
- m means length of affinityTerms.

But of course we need O(n•m) space.

@bsalamat do you think it's a viable approach?

Reply:
At this point, I feel this feature is not worth the complexity given that we have not seen any strong desire from users to have it. When we have the scheduling framework, this can be easily implemented as a post-filter plugin. Once all filters (predicates) are executed, we check:
1. Pod is unschedulable.
2. Pod has affinity to itself.
3. Find nodes that have failed "inter-pod affinity" only.
4. Mark those nodes feasible for the pod.

Reply:
Well, the above algorithm is correct, only if we run all predicates for each node even after some of the predicates have failed.

Reply:
> When we have the scheduling framework, this can be easily implemented as a post-filter plugin.

That makes more sense - it's not a good deal to make it as default behavior with such cost.

> Well, the above algorithm is correct, only if we run all predicates for each node even after some of the predicates have failed.

Yes, it's not a "on-demand calculation" algorithm.

Reply:
So @bsalamat are you ok with proceeding with (1) discard recent 2 commits which are for the special PodAffinity-match-itself case, (2) remove check `targetPodMatchesAffinityOfPod(pod, pod))` as the behavior is already incorrect

Reply:
Thanks I got it. Let me do a summary here:

Right now, in terms of supporting PodAffinity on multiple nodes, current PR implements most cases. So we can keep relative commits as is.

Regarding the case support podAffinity to self. Here is current status and possible way to proceed without new scheduling framework:

1. I'm in halfway of implementing the idea described in https://github.com/kubernetes/kubernetes/pull/68725#discussion_r222424295. And as you pointed out, it's an algorithm happening in calculating/updating PredicateMetadata phase. In other words, it's calculated against all nodes in advance.

2. Another way is to use chan/sync.cond in InterPodPredicate phase. Say to only calculate against nodes which have passed previous predicates.

@bsalamat for solution 1, by considering the time/space complexity and behavior, I think your take is to stop investing on it, right?

If so, based on fact that scheduling framework won't come too soon, I'd give a try on solution 2, but will lower the priority in my list. Sounds good?

Reply:
Yes to postponing solution 1, but your solution 2 is not clear to me. Can you elaborate on it? Please keep in mind that matching namespace and selector of inter-pod affinity should be done for all pods in a cluster, not only for pods on nodes that have passed previous predicates.

Reply:
> your solution 2 is not clear to me. Can you elaborate on it?

Right now it's a rough idea - pass in a shared chan/sync.cond into PodAffinityChecker, so that for a check (on one node), it can communicate with other checks: broadcast() if it failed on podAffinity check specifically, or close the chan if it passed, so that other failed checkers don't need to wait for a (possible) eventual all-failure signal.

> Please keep in mind that matching namespace and selector of inter-pod affinity should be done for all pods in a cluster

Sure, topologyKeyMaps are calculated/synced in PredicateMetadata phase, and they are always calculated against all pods. It's still the global thing.

Let me give an example to describe expected behavior in solution 2:

```
- Assume that the cluster has two nodes: 
- nodeA located in in zone1/region1, and has 4 allocatable cpus
- nodeB located in zone2/region1, and has 2 allocatable cpus,

- There are two existing pods on these nodes:
- Pod1:
    - nodeName: "nodeA"
    - no labels
- Pod2:
    - nodeName: "nodeB"
    - label "foo":"", "bar":""

- Pod3 comes in with inter-pod affinity:
    - affinity terms:
        - {label "foo" exists, topologyKey: "region"}
        - {label "bar" exists, topologyKey: "zone"}
    - label "foo":"", "bar":""

- case I: if pod3 request <= 2 cpus, it can fit on nodeB, and cannot fit on nodeA
- case II: if pod3 request > 2 cpus, it cannot fit on nodeB due to not enough cpu in nodeB, and
it can fit on nodeA as pod3 matches itself. Unlike solution 1, for this case whether nodeB can fit is not respected/calculated any more as it failed in previous predicate.
```

Reply:
Per discussion with @bsalamat in today's sig meeting, solution 2 changes the behavior what we have today. Say in above example, even if pod3 failed on nodeB due to insufficient cpu, nodeB still need to be considered to answer "whether incoming pod can fit on a node".

So I will go back to solution 1, but lower the priority.